### DISCUSSION TITLE
How to Add Global Web API Polyfills in App Router

### DISCUSSION THREAD

**Discussion by xiefenga** [vercel/next.js#73730] [App Router]
Hi,  

I'm using Next.js with the App Router and need to add global web API polyfills to ensure compatibility with older browsers. For example, I want to polyfill `structuredClone` using the `@ungap/structured-clone` package.  

Here’s what I’ve tried:  

### Polyfill Setup  

```ts
// lib/polyfills.ts
import structuredClone from "@ungap/structured-clone";

export const applyPolyfills = () => {
  if (!("structuredClone" in globalThis)) {
    console.log("structuredClone not found, attaching polyfill");
    // @ts-ignore
    globalThis.structuredClone = structuredClone;
  }
};
```  

### Client Component  

```tsx
'use client';

import { useEffect, useState } from 'react';
import { applyPolyfills } from '../lib/polyfills';

export const ClientPolyfills = ({ children }: React.PropsWithChildren) => {
  const [isMounted, setIsMounted] = useState(false);

  useEffect(() => {
    applyPolyfills();
    setIsMounted(true);
  }, []);

  return isMounted ? children : null;
};
```  

### Usage in Root Layout  

```tsx
const RootLayout = ({ children }: React.PropsWithChildren) => {
  return (
    <html lang="zh">
      <body>
        <ClientPolyfills>
          <Providers>{children}</Providers>
        </ClientPolyfills>
      </body>
    </html>
  );
};
```  

This approach works but has several issues:  

1. **Direct calls to `structuredClone` in modules fail.**  
   If a module directly calls `structuredClone` before the polyfill is applied, it throws an error since the polyfill isn't globally available at the time of execution.  

2. **Rendering delay.**  
   The layout's children rendering is delayed until the `ClientPolyfills` component has mounted and applied the polyfills. This isn't ideal, especially for critical content.  

3. **Feels excessive.**  
   Wrapping the entire layout for a global task like polyfilling doesn’t seem efficient.  

4. **Lack of clarity.**  
   Managing polyfills in a client-side component doesn't feel like the right place for a global setup.  

### Question  

How can I globally and safely add web API polyfills like `structuredClone` in Next.js with the App Router, ensuring:  
- The polyfill is applied early enough to avoid errors in modules that call the function directly.  
- It's efficient and doesn't delay rendering unnecessarily.  

Is there a recommended way to handle this in the App Router?  

Thank you in advance for any insights or solutions!

**Comment by icyJoseph** [comment]
It used to be possible to import the polyfill, with the use client directive, on top, into layout, but that's not working on recent versions... 

I'll try to investigate more, though I'm short on time these days.

**Comment by kuttymbekov** [comment]
any updates?

**Reply by icyJoseph** [reply]
https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client#polyfills
